### PR TITLE
Use readers-writer lock for page table nodes

### PR DESCRIPTION
**Pull request by junyang-zh**
As pointed out in #1395 , the PML4 lock attributes the major contention in SMP for page faults. However the PML4 page table is rarely written to. A read-write lock is better for the scalability.

### CODE REVIEW

File: ostd/src/mm/page_table/node/entry.rs [hunk 1]
```diff
@@ -26,14 +26,54 @@ where
...
+impl<'a, E: PageTableEntryTrait, C: PagingConstsTrait> Entry<'a, E, C>
+where
+    [(); C::NR_LEVELS as usize]:,
+{
+    /// Returns if the entry maps to a page table node.
+    pub(in crate::mm) fn is_node(&self) -> bool {
+        self.pte.is_present() && !self.pte.is_last(self.node.level())
+    }
+
+    /// Gets a owned handle to the child.
```
Comment (cchanging) [inline ostd/src/mm/page_table/node/entry.rs:43]: -> an owned

File: ostd/src/mm/page_table/mod.rs [hunk 2]
```diff
@@ -100,17 +100,17 @@ impl PageTable<KernelMode> {
...
+        let mut root_node = self.root.clone_shallow().lock_write();
         let mut new_node =
             PageTableNode::alloc(PagingConsts::NR_LEVELS, MapTrackingStatus::NotApplicable);
 
         // Make a shallow copy of the root node in the kernel space range.
         // The user space range is not copied.
         const NR_PTES_PER_NODE: usize = nr_subpage_per_huge::<PagingConsts>();
         for i in NR_PTES_PER_NODE / 2..NR_PTES_PER_NODE {
-            let root_entry = root_node.entry(i);
+            let root_entry = root_node.entry_mut(i);
```
Comment (cchanging) [inline ostd/src/mm/page_table/mod.rs:111]: Why need `entry_mut` here?

File: ostd/src/mm/page_table/cursor.rs [hunk 3]
```diff
@@ -225,7 +224,7 @@ where
 
             match self.cur_entry().to_owned() {
```
Comment (cchanging) [inline ostd/src/mm/page_table/cursor.rs:225]: The internal function `cur_entry` gets an `EntryMut`, is there a scenario where `cur_entry` does not require mut ( it seems like here)? I think an additional `cur_entry_mut` could be provided, or directly offer `cur_node` and use `entry`/`entry_mut` following it.

File: ostd/src/mm/page/meta.rs [hunk 4]
```diff
@@ -221,15 +221,49 @@ impl<E: PageTableEntryTrait, C: PagingConstsTrait> PageTablePageMeta<E, C>
...
             nr_children: UnsafeCell::new(0),
             level,
             is_tracked,
-            lock: AtomicU8::new(1),
+            lock: AtomicU32::new(Self::WRITE_LOCK_BIT),
             _phantom: PhantomData,
         }
     }
+
+    pub(crate) fn lock_read(&self) {
```
Comment (cchanging) [inline ostd/src/mm/page/meta.rs:236]: Here provides additional implementation logics for `RwLock`. I was thinking that if we could add `acquire_read(write)_lock` and `release_lock` methods to `RwLock` (similar to the `SpinLock`), and make them public in the pt module. Then, the original `RwLock` implementation could be based on these APIs to reduce code duplication and we could directly store an `RwLock<()>` here.

This is not directly related to this PR and we can address it later if you think it is reasonable (to reduce TCB).

File: ostd/src/mm/page_table/node/entry.rs [hunk 5]
```diff
@@ -26,14 +26,54 @@ where
...
+    pub(in crate::mm) fn to_owned(&self) -> Child<E, C> {
+        // SAFETY: The entry structure represents an existent entry with the
+        // right node information.
+        unsafe { Child::clone_from_pte(&self.pte, self.node.level(), self.node.is_tracked()) }
+    }
+
+    /// Create a new entry at the node.
```
Comment (cchanging) [inline ostd/src/mm/page_table/node/entry.rs:50]: -> Creates

**Review by cchanging — COMMENTED** [review]
The changes in this PR are reasonable and can significantly optimize parallel performance!

**Comment by junyang-zh** [comment]
I think this PR is not the correct way to resolve the contention. I'll post other update PRs.

**Pull request closed without merging**